src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use Scope;
+
+/// Given ordering for the success case in a compare-exchange operation, returns the strongest
+/// appropriate ordering for the failure case.
+#[inline]
+fn strongest_failure_ordering(ord: Ordering) -> Ordering {
```

Comment:
Personally, I tend to avoid `compare_exchange` because:

1. It's so verbose.
2. `compare_and_swap` with `strongest_failure_ordering` usually does exactly what I want anyway.
3. I'm slightly afraid of screwing up the two orderings (choosing one ordering is difficult enough already :)).

We could add 4 new methods: `compare_exchange{,_weak,_owned,_weak_owned}`, although I'd prefer not adding them right now. But if someone really needs them in the future, I won't object to their addition. How does that sound? :)